books/RayTracingTheNextWeek.html

```diff
@@ -2904,24 +2951,112 @@
-to that.
+For a simple operation like a translation, thinking of the operation as a simple movement of the
+initial ray is a fine way to reason about what's going on. But, for a more complex operation like a
+rotation, it can be easy to accidentally get your terms crossed (or forget a negative sign), so it's
+better to consider a rotation as a change of coordinates. The pseudocode for the `translate::hit`
+function above describes the function in terms of _moving_:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+        // Move the ray backwards by the offset
+
+        // Determine where (if any) an intersection occurs along the offset ray
+
+        // Move the intersection point forwards by the offset
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [translate-hit-move]: <kbd>[hittable.h]</kbd> The moving translate hit code]
+
+But this can also be thought of in terms of a _changing of coordinates_:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+        // Change the ray from world space to object space
+
+        // Determine where (if any) an intersection occurs in object space
+
+        // Change the intersection point from object space to world space
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [translate-hit-coord]: <kbd>[hittable.h]</kbd> The change of coordinates translate hit code]
 
 <div class='together'>
+Rotating an object will not only change the point of intersection, but will also change the surface
+normal vector, which will change the direction of reflections and refractions. So we need to change the
+normal as well. Fortuneatly, we can just treat the normal like a vector, and so use the same
+formulas as above.
+
 For a y-rotation class we have:
 
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
     class rotate_y : public hittable {
       public:
-        rotate_y(shared_ptr<hittable> p, double angle) : object(p) {
+        ...
+
+        bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+            // Change the ray from world space to object space
+            auto origin = r.origin();
+            auto direction = r.direction();
+
+            origin[0] = cos_theta*r.origin()[0] - sin_theta*r.origin()[2];
+            origin[2] = sin_theta*r.origin()[0] + cos_theta*r.origin()[2];
+
+            direction[0] = cos_theta*r.direction()[0] - sin_theta*r.direction()[2];
+            direction[2] = sin_theta*r.direction()[0] + cos_theta*r.direction()[2];
+
+            ray rotated_r(origin, direction, r.time());
```

Comment:
Getting more convinced than ever that we should strongly think about implementing ray and vector transforms. This function is littered with minutia about vector, point and ray transforms that fuzzes the higher-level operations happening here. I believe this would make the codebase more clear and more compact.

Reply:
BTW, with our point/vector duality inherent in the `vec3` class, we'd need

- translate_as_point()
- translate_as_vector() (or just omit this entirely as it's a no-op)
- scale()
- scale_as_normal()
- rotate()

And I'd almost rather have a `rotate(const vec3& axis, double degrees)` method than `rotate_x()`, `rotate_y()`, `rotate_z()` methods, but `¯\_(ツ)_/¯`.

Reply:
Thankfully, the `ray` class would just have `translate()`, `rotate()` and `scale()` methods.

Reply:
Nothing actionable in this PR.